sql/core/src/main/scala/org/apache/spark/sql/execution/systemcatalog/InformationSchema.scala

```diff
@@ -0,0 +1,336 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.systemcatalog
+
+import java.sql.{Date, Timestamp}
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.commons.lang3.StringUtils
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.expressions.Alias
+import org.apache.spark.sql.catalyst.plans.logical.Project
+import org.apache.spark.sql.execution.datasources._
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+
+/**
+ * INFORMATION_SCHEMA is a database consisting views which provide information about all of the
+ * tables, views, columns in a database.
+ */
+object InformationSchema {
+  val INFORMATION_SCHEMA = "information_schema"
+  /**
+   * Register INFORMATION_SCHEMA database.
+   */
+  def registerInformationSchema(sparkSession: SparkSession): Unit = {
+    sparkSession.sql(s"CREATE DATABASE IF NOT EXISTS $INFORMATION_SCHEMA")
+    registerView(sparkSession, new DatabasesRelationProvider, Seq("schemata", "databases"))
+    registerView(sparkSession, new TablesRelationProvider, Seq("tables"))
+    registerView(sparkSession, new ViewsRelationProvider, Seq("views"))
+    registerView(sparkSession, new ColumnsRelationProvider, Seq("columns"))
+    registerView(sparkSession, new SessionVariablesRelationProvider, Seq("session_variables"))
+  }
+
+  /**
+   * Register a INFORMATION_SCHEMA relation provider as a temporary view of Spark Catalog.
+   */
+  private def registerView(
+      sparkSession: SparkSession,
+      relationProvider: SchemaRelationProvider,
+      names: Seq[String]) {
+    val plan =
+      LogicalRelation(relationProvider.createRelation(sparkSession.sqlContext, null, null)).analyze
+    val projectList = plan.output.zip(plan.schema).map {
+      case (attr, col) => Alias(attr, col.name)()
+    }
+    sparkSession.sessionState.executePlan(Project(projectList, plan))
+    for (name <- names) {
+      // TODO(dongjoon): This is a hack to give a database concept for Spark temporary views.
+      // We should generalize this later.
+      sparkSession.sessionState.catalog.createTempView(s"$INFORMATION_SCHEMA.$name",
+        plan, overrideIfExists = true)
+    }
+  }
+
+  /**
+   * Compile filter array into single string condition.
+   */
+  private[systemcatalog] def getConditionExpressionString(filters: Array[Filter]): String = {
+    val str = filters.flatMap(InformationSchema.compileFilter).map(p => s"($p)").mkString(" AND ")
+    if (str.length == 0) "TRUE" else str
+  }
+
+  /**
+   * Convert filter into string expression.
+   */
+  private[systemcatalog] def compileFilter(f: Filter): Option[String] = {
```

Comment:
Oh I see. Then if we won't dedup this for now, let's leave some comments saying if we should change some piece of code please don't forget to change the other? What do you think?